core/shared/src/main/scala/fs2/AsyncPull.scala

```diff
@@ -35,22 +35,20 @@ sealed abstract class AsyncPull[F[_],A] { self =>
   def race[B](b: AsyncPull[F,B])(implicit F: Effect[F], ec: ExecutionContext): AsyncPull[F,Either[A,B]] = new AsyncPull[F, Either[A,B]] {
     def get = cancellableGet.flatMap(_._1)
     def cancellableGet = FreeC.Eval(for {
-      promise <- async.promise[F,Either[Throwable, Either[A,B]]]
+      promise <- async.promise[F, Either[Throwable, Either[A,B]]]
       t0 <- self.cancellableGet.run
       (a, cancelA) = t0
       t1 <- b.cancellableGet.run
       (b, cancelB) = t1
-      fa = a.run.map(Left(_): Either[A, B])
-      fb = b.run.map(Right(_): Either[A, B])
-      _ <-  async.fork(fa.attempt.flatMap(x => promise.setSync(x)))
-      _ <-  async.fork(fb.attempt.flatMap(x => promise.setSync(x)))
-    } yield {
-      (FreeC.Eval(promise.get.flatMap {
-        case Left(e) => F.raiseError[Either[A, B]](e)
-        case Right(Left(a)) => cancelB.run.as(Left(a): Either[A, B])
-        case Right(Right(b)) => cancelA.run.as(Right(b): Either[A, B])
-      }), FreeC.Eval(cancelA.run *> cancelB.run))
-    })
+      fa = a.run.map(Left.apply).attempt
+      fb = b.run.map(Right.apply).attempt
+      _ <-  async.fork {
+        fa.flatMap(x => promise.complete(x).handleErrorWith(_ => cancelA.run))
```

Comment:
I think `cancelA` and `cancelB` are swapped on lines 46/47?

Reply:
They are, and that's by design.
Since now completing on a promise gives an error if the promise is already set, we can now now if we arrived second to the race, and cancel _ourselves_. Before, we had to check the result and cancel the _other_ runner.  wdyt?

Reply:
Hm I see. Wouldn't that delay cancelation until after the `complete` has occurred, which is after the action has finished?

Reply:
Or I guess I can `flatMap` on `complete` and cancel the other `F`

Reply:
Yeah let's do that instead of reverting

Reply:
:+1: